File: frontend/services/reportService.ts
import { api } from "@/lib/api/client";
import { SoilHealthReport } from "@/types/soilHealth";

export interface ReportVerificationResult {
  verified: boolean;
  status: string;
  report_no?: string;
  receipt_no?: string;
  farmer_name?: string;
  gat_no?: string;
  village?: string;
  taluka?: string;
  district?: string;
  report_date?: string;
  sample_date?: string;
  is_demo?: boolean;
  laboratory_name?: string;
}

export const reportService = {
  /**
   * Fetch complete Soil Health Report metadata and parameters.
   */
  async getFieldReport(fieldId: string | number): Promise<SoilHealthReport> {
    const response = await api.get<SoilHealthReport>(`/reports/${fieldId}`);
    return response.data;
  },

  /**
   * Fetch Soil Health Card specific diagnostics.
   */
  async getSoilHealthCard(fieldId: string | number): Promise<SoilHealthReport> {
    const response = await api.get<SoilHealthReport>(`/reports/${fieldId}/soil-health-card`);
    return response.data;
  },

  /**
   * Trigger download of official Soil Health Card PDF from backend.
   */
  async downloadSoilHealthCardPdf(fieldId: string | number, lang: string = "en"): Promise<void> {
    const response = await api.get(`/reports/${fieldId}/soil-health-card/pdf`, {
      params: { lang },
      responseType: "blob",
    });

    let filename = `SoilPilot_Soil_Health_Card_Gat_${fieldId}.pdf`;
    const disposition = response.headers["content-disposition"];
    if (disposition && disposition.indexOf("filename=") !== -1) {
      const matches = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/.exec(disposition);
      if (matches != null && matches[1]) {
        filename = matches[1].replace(/['"]/g, "");
      }
    }

    const blob = new Blob([response.data], { type: "application/pdf" });
    const blobUrl = window.URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = blobUrl;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(blobUrl);
  },

  /**
   * Trigger download of official Detailed Soil Report PDF from backend.
   */
  async downloadDetailedReportPdf(fieldId: string | number, lang: string = "en"): Promise<void> {
    const response = await api.get(`/reports/${fieldId}/detailed/pdf`, {
      params: { lang },
      responseType: "blob",
    });

    let filename = `SoilPilot_Detailed_Soil_Report_Gat_${fieldId}.pdf`;
    const disposition = response.headers["content-disposition"];
    if (disposition && disposition.indexOf("filename=") !== -1) {
      const matches = /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/.exec(disposition);
      if (matches != null && matches[1]) {
        filename = matches[1].replace(/['"]/g, "");
      }
    }

    const blob = new Blob([response.data], { type: "application/pdf" });
    const blobUrl = window.URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = blobUrl;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(blobUrl);
  },

  /**
   * Verify authenticity of a report number.
   */
  async verifyReport(reportNo: string): Promise<ReportVerificationResult> {
    const response = await api.get<ReportVerificationResult>(`/reports/verify/${encodeURIComponent(reportNo)}`);
    return response.data;
  },
};

File: backend/app/services/report_service.py
"""Reports Service for SoilPilot Phase 9.
Handles report data retrieval, verification, authorization checks, and PDF exports.
Single source of truth between online preview and PDF documents.
"""
from typing import Dict, Any, Optional
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from app.models.field import Field
from app.models.farmer import Farmer
from app.models.soil_health import SoilReport
from app.services.soil_health_service import soil_health_service
from app.services.pdf_service import pdf_service


def sanitize_filename_part(text: str) -> str:
    """Sanitize string for clean, ASCII-safe HTTP header filenames."""
    clean = "".join(c if (c.isascii() and c.isalnum()) or c in ("-", "_") else "_" for c in text.strip())
    while "__" in clean:
        clean = clean.replace("__", "_")
    return clean.strip("_")


class ReportService:
    """Domain service for reports and PDF exports."""

    def get_report_data(self, db: Session, field_identifier: str) -> Dict[str, Any]:
        """Fetch unified soil report data."""
        data = soil_health_service.get_report_for_field(db, field_identifier)
        if not data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Field with ID or Gat '{field_identifier}' not found.",
            )
        return data

    def verify_field_access(
        self,
        db: Session,
        field_identifier: str,
        token_payload: Optional[Dict[str, Any]],
    ) -> None:
        """
        Verify that the requested field belongs to the current farmer's session.
        Permits access if:
          1. Field identifier matches session token's field_id or gat_no
          2. Field is demo field (Gat 104) for platform evaluation
          3. Token belongs to the farmer who owns the field
        """
        norm_id = str(field_identifier).strip().lower()
        if norm_id in ["demo-field-gat-104", "demo", "gat-104", "104"]:
            return  # Demo field is always accessible for review

        # If session token is provided, enforce ownership
        if token_payload:
            sess_field_id = token_payload.get("field_id")
            sess_gat = str(token_payload.get("gat_no", "")).strip().lower()
            sess_farmer_id = token_payload.get("farmer_id")

            if str(sess_field_id) == str(field_identifier) or sess_gat == norm_id:
                return

            # Check if farmer owns this field
            if sess_farmer_id:
                target_field = soil_health_service._find_field(db, field_identifier)
                if target_field and target_field.farmer_id == sess_farmer_id:
                    return

            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Unauthorized: You do not have permission to access reports for this field.",
            )

    def generate_soil_health_card_pdf(
        self,
        db: Session,
        field_identifier: str,
        lang: str = "en",
        base_url: str = "https://soilpilot.gov.in",
    ) -> tuple[bytes, str]:
        """Generates Soil Health Card PDF binary and clean filename."""
        data = self.get_report_data(db, field_identifier)
        report_meta = data.get("report") or {}
        report_no = report_meta.get("report_no", "SPL/2026/SL-0104")
        verify_url = f"{base_url}/reports?verify={report_no}"

        pdf_bytes = pdf_service.generate_soil_health_card_pdf(
            report_data=data,
            lang=lang,
            verify_url=verify_url,
        )

        gat_no = sanitize_filename_part(str(data.get("field", {}).get("gat_no", "104")))
        farmer_name = sanitize_filename_part(str(data.get("farmer", {}).get("name", "")))
        if farmer_name and farmer_name.lower() != "farmer":
            filename = f"SoilPilot_Soil_Health_Card_{farmer_name}_Gat_{gat_no}.pdf"
        else:
            filename = f"SoilPilot_Soil_Health_Card_Gat_{gat_no}.pdf"

        return pdf_bytes, filename

    def generate_detailed_soil_report_pdf(
        self,
        db: Session,
        field_identifier: str,
        lang: str = "en",
        base_url: str = "https://soilpilot.gov.in",
    ) -> tuple[bytes, str]:
        """Generates Detailed Soil Report PDF binary and clean filename."""
        data = self.get_report_data(db, field_identifier)
        report_meta = data.get("report") or {}
        report_no = report_meta.get("report_no", "SPL/2026/SL-0104")
        verify_url = f"{base_url}/reports?verify={report_no}"

        pdf_bytes = pdf_service.generate_detailed_soil_report_pdf(
            report_data=data,
            lang=lang,
            verify_url=verify_url,
        )

        gat_no = sanitize_filename_part(str(data.get("field", {}).get("gat_no", "104")))
        farmer_name = sanitize_filename_part(str(data.get("farmer", {}).get("name", "")))
        if farmer_name and farmer_name.lower() != "farmer":
            filename = f"SoilPilot_Detailed_Soil_Report_{farmer_name}_Gat_{gat_no}.pdf"
        else:
            filename = f"SoilPilot_Detailed_Soil_Report_Gat_{gat_no}.pdf"

        return pdf_bytes, filename

    def verify_report(self, db: Session, report_no: str) -> Dict[str, Any]:
        """Verify authenticity of a report number for QR scan."""
        clean_no = report_no.strip()
        db_report = db.query(SoilReport).filter(SoilReport.report_no.ilike(clean_no)).first()

        if db_report:
            field = db_report.field
            farmer = field.farmer if field else None
            village = field.village if field else None
            taluka = village.taluka if village else None
            district = taluka.district if taluka else None

            return {
                "verified": True,
                "status": "Authentic Certified Soil Test Record",
                "report_no": db_report.report_no,
                "receipt_no": db_report.receipt_no,
                "farmer_name": farmer.full_name if farmer else "Farmer",
                "gat_no": field.gat_no if field else "N/A",
                "village": village.name if village else "N/A",
                "taluka": taluka.name if taluka else "N/A",
                "district": district.name if district else "Pune",
                "report_date": db_report.report_date,
                "sample_date": db_report.sample_date,
                "is_demo": db_report.is_demo,
                "laboratory_name": db_report.laboratory_name,
            }

        # Check for demo report number
        if clean_no.upper() in ["SPL/2026/SL-0104", "SPL-2026-0104", "DEMO-104"]:
            return {
                "verified": True,
                "status": "Authentic Demonstration Soil Test Record",
                "report_no": "SPL/2026/SL-0104",
                "receipt_no": "REC-7842/26",
                "farmer_name": "Pradip Bhauso Shelar",
                "gat_no": "104",
                "village": "Malegaon Bk",
                "taluka": "Baramati",
                "district": "Pune",
                "report_date": "20-09-2026",
                "sample_date": "15-09-2026",
                "is_demo": True,
                "laboratory_name": "SoilPilot Soil Testing & Diagnostic Laboratory",
            }

        return {
            "verified": False,
            "status": "Report Not Found in Cadastral Database",
            "report_no": clean_no,
        }


report_service = ReportService()
